server/ladder_service.py

```diff
@@ -125,64 +151,77 @@ def _cancel_existing_searches(self, initiator: Player) -> List[Search]:
+            game.set_player_option(host.id, 'Color', 1)
+            game.set_player_option(guest.id, 'Color', 2)
+            game.set_player_option(host.id, 'Army', 2)
+            game.set_player_option(guest.id, 'Army', 3)
+
+            # Remembering that "Team 1" corresponds to "-": the non-team.
+            game.set_player_option(host.id, 'Team', 1)
+            game.set_player_option(guest.id, 'Team', 1)
+
+            mapname = map_path[5:-4]
+            # FIXME: Database filenames contain the maps/ prefix and .zip suffix.
+            # Really in the future, just send a better description
+            self._logger.debug("Starting ladder game: %s", game)
+            await host.lobby_connection.launch_game(game, is_host=True, use_map=mapname)
+            try:
+                hosted = await game.await_hosted()
+                if not hosted:
+                    raise TimeoutError("Host left lobby")
+            except TimeoutError:
+                msg = {"command": "game_launch_cancelled"}
+                with contextlib.suppress(DisconnectedError):
+                    # Sending to host might not be needed, but we do it anyways
+                    await asyncio.gather(
+                        host.send_message(msg),
+                        guest.send_message(msg)
+                    )
+                # TODO: Uncomment this line once the client supports `game_launch_cancelled`.
+                # Until then, returning here will cause the client to think it is
+                # searching for ladder, even though the server has already removed it
+                # from the queue.
+
+                # return
+                self._logger.debug("Ladder game failed to launch due to a timeout")
+
+            # TODO: Graceful handling of NoneType errors due to disconnect
+            await guest.lobby_connection.launch_game(
+                game, is_host=False, use_map=mapname
             )
-            # TODO: Uncomment this line once the client supports `game_launch_timeout`.
-            # Until then, returning here will cause the client to think it is
-            # searching for ladder, even though the server has already removed it
-            # from the queue.
-            # return
-            self._logger.debug("Ladder game failed to launch due to a timeout")
-
-        await guest.lobby_connection.launch_game(
-            game, is_host=False, use_map=mapname
-        )
-        self._logger.debug("Ladder game launched successfully")
+            self._logger.debug("Ladder game launched successfully")
+        except Exception:
+            self._logger.exception("Failed to start ladder game!")
+            msg = {"command": "game_launch_cancelled"}
+            with contextlib.suppress(DisconnectedError):
+                await asyncio.gather(
+                    host.send_message(msg),
+                    guest.send_message(msg)
+                )
```

Comment:
I'm thinking about that big 60 lines of code inner content, that is in try. It's big and it's not readable much, i know, that there is a lot of small operations. Maybe we could extract it to separate function, to see the final logic: 
```python3
self._logger.debug("Starting ladder game: %s", game)
            await host.lobby_connection.launch_game(game, is_host=True, use_map=mapname)
            try:
                hosted = await game.await_hosted()
                if not hosted:
                    raise TimeoutError("Host left lobby")
            except TimeoutError:
                msg = {"command": "game_launch_cancelled"}
                with contextlib.suppress(DisconnectedError):
                    # Sending to host might not be needed, but we do it anyways
                    await asyncio.gather(
                        host.send_message(msg),
                        guest.send_message(msg)
                    )
                # TODO: Uncomment this line once the client supports `game_launch_cancelled`.
                # Until then, returning here will cause the client to think it is
                # searching for ladder, even though the server has already removed it
                # from the queue.

                # return
                self._logger.debug("Ladder game failed to launch due to a timeout")

            # TODO: Graceful handling of NoneType errors due to disconnect
            await guest.lobby_connection.launch_game(
                game, is_host=False, use_map=mapname
            )
```

Like: `prepare_game`, `connect_users` (`host` + `guest`)
\+ I think, we can get rid of inner `except TimeoutError:`. This is a little bit ridiculous to have variable `hosted` and then raise, catch, and then again... this is useless load for our server. i know, that `python way`, etc.. But generating exception still has to collect whole stack-trace, and gather other information, run raise mechanics etc.. instead of boolean check...

```python3
            hosted = await game.await_hosted()
            if not hosted:
                ....
```

Reply:
Problem is that `await_hosted()` can raise `TimeoutError` too. This way folds the error handling neatly into one case, instead of having to write the code twice.